src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_agents.clj

```diff
@@ -86,87 +87,102 @@
    pool :- jruby-schemas/pool-queue-type]
   (= expected-pool-size (count (.getRegisteredElements pool))))
 
-(schema/defn ^:always-validate
-  swap-and-drain-pool!
-  "Replace the current pool with a new pool and drain the old pool,
-  optionally refilling the new pool with fresh jrubies."
+(schema/defn collect-all-jrubies :- [JRubyInstance]
+  "Locks the pool and borrows all the instances"
+  [pool-state :- jruby-schemas/PoolState]
+  (let [pool-size (:size pool-state)
+        pool (:pool pool-state)
+        borrow-fn (fn [_] (jruby-internal/borrow-from-pool!*
+                           jruby-internal/borrow-without-timeout-fn
+                           pool))]
+    (.lock pool)
+    (try
+      (mapv borrow-fn (range pool-size))
+      (catch Exception e
+        (.clear pool)
+        (.insertPill pool (PoisonPill. e))
+        (throw (IllegalStateException.
+                "There was a problem borrowing a JRubyInstance from the pool."
+                e)))
+      (finally
+        (.unlock pool)))))
+
+(schema/defn cleanup-and-refill-pool
+  "Cleans up the given instances and optionally refills the pool with
+  new instances. Should only be called from the modify-instance-agent"
   [pool-context :- jruby-schemas/PoolContext
-   old-pool-state :- jruby-schemas/PoolState
-   new-pool-state :- jruby-schemas/PoolState
+   pool-state :- jruby-schemas/PoolState
+   old-instances :- [JRubyInstance]
    refill? :- schema/Bool]
-  (let [{:keys [config]} pool-context
-        pool-state-atom (jruby-internal/get-pool-state-container pool-context)
-        new-pool (:pool new-pool-state)
-        old-pool (:pool old-pool-state)
-        old-pool-size (:size old-pool-state)
+  (let [pool (:pool pool-state)
+        pool-size (:size pool-state)
+        new-instance-ids (map inc (range pool-size))
```

Comment:
I think `next-instance-id`'s current implementation might not work for this because it just tried to determine a new id based on the current id + the size of the current pool.  That, I think, generally worked for the old pool implementation because we only had individual instance flushes and not full pool flushes to deal with for the same pool - since the full pool ones always resulted in the creation of a new pool.  I think it would be more useful from the perspective of the log messages overall to just assign an id that is more guaranteed to be unique per pool instance over the life of a puppetserver service run.